### DISCUSSION TITLE
Workspace packages peerDependencies don't work

### DISCUSSION THREAD

**Discussion by MarcioMeier** [pnpm/pnpm#16186] [General]
### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### pnpm version

9.1.4

### Which area(s) of pnpm are affected? (leave empty if unsure)

Lockfile

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/MarcioMeier/pnpm-workspace-peer-dep-resolution-bug

### Reproduction steps

Create 2 packages in the pnpm workspace.

In the package `a` add a library as peer dependency.

In the package `b` add the same library in another version than the specified in package `a`

It will cause the library version miss match.

### Describe the Bug

The packages managed by the pnpm workspace don't declare its peer dependencies in the lockfile like a published library. This issue make impossible to use peer dependencies in the managed pnpm workspace packages.

### Expected Behavior

I expect to the workspace package's peer dependencies to work just like the published libraries.

### Which Node.js version are you using?

20.9.0

### Which operating systems have you used?

- [X] macOS
- [ ] Windows
- [ ] Linux

**Comment by LJJCherry** [comment]
I've met the same question

**Comment by zkochan** [comment]
A pull request has been opened to address this: https://github.com/pnpm/pnpm/pull/15448

### Approach
- In TypeScript v11 (`@pnpm/installing.deps-resolver` and `@pnpm/deps.inspection.peers-checker`), inspect consuming workspace projects and fall back to the workspace root for linked packages declaring peer dependencies.
- In Rust v12 (`pnpm-deps-inspection-peers`), resolve peer providers from consuming importers and the workspace root fallback.
- Added regression tests verifying resolution of workspace package peer dependencies.

---
*Generated by Antigravity (Agent 1)*

**Comment by zkochan** [comment]
This one is not a bug. It works as expected. The right way to handle this is probably using version catalogs, or installing the peers in the root of the repo. If you need multiple versions of a peer your only option is using injected dependencies.
